scripts/pylib/twister/twisterlib/testplan.py

```diff
@@ -842,32 +842,35 @@ def apply_filters(self, **kwargs):
                 #
-                # A simple example is keying on arch and simulation to run a test once per unique (arch, simulation) platform.
+                # A simple example is keying on arch and simulation
+                # to run a test once per unique (arch, simulation) platform.
                 if not ignore_platform_key and hasattr(ts, 'platform_key') and len(ts.platform_key) > 0:
-                    # form a key by sorting the key fields first, then fetching the key fields from plat if they exist
-                    # if a field does not exist the test is still scheduled on that platform as its undeterminable.
                     key_fields = sorted(set(ts.platform_key))
-                    key = [getattr(plat, key_field) for key_field in key_fields]
-                    has_all_fields = True
-                    for key_field in key_fields:
-                        if key_field is None or key_field == 'na':
-                            has_all_fields = False
-                    if has_all_fields:
-                        test_key = copy.deepcopy(key)
-                        test_key.append(ts.name)
-                        test_key = tuple(test_key)
-                        keyed_test = keyed_tests.get(test_key)
+                    keys = [getattr(plat, key_field) for key_field in key_fields]
+                    for key in keys:
+                        if key is None or key == 'na':
+                            instance.add_filter(
+                                f"Excluded platform missing key fields demanded by test {key_fields}",
+                                Filters.PLATFORM
+                            )
+                            break
+                    else:
+                        test_keys = copy.deepcopy(keys)
+                        test_keys.append(ts.name)
+                        test_keys = tuple(test_keys)
+                        keyed_test = keyed_tests.get(test_keys)
                         if keyed_test is not None:
                             plat_key = {key_field: getattr(keyed_test['plat'], key_field) for key_field in key_fields}
-                            instance.add_filter(f"Excluded test already covered for key {tuple(key)} by platform {keyed_test['plat'].name} having key {plat_key}", Filters.PLATFORM_KEY)
+                            instance.add_filter(
+                                f"Excluded test already covered for key {tuple(key)} by platform {keyed_test['plat'].name} having key {plat_key}",
+                                Filters.PLATFORM_KEY
+                            )
                         else:
-                            keyed_tests[test_key] = {'plat': plat, 'ts': ts}
-                    else:
-                        instance.add_filter(f"Excluded platform missing key fields demanded by test {key_fields}", Filters.PLATFORM)
+                            keyed_tests[test_keys] = {'plat': plat, 'ts': ts}
```

Comment:
an issue that existed in original implementation: we need to check if that platform was filtered and should not use it if it was filtered, so 

```
-                            keyed_tests[test_key] = {'plat': plat, 'ts': ts}
+                            if not instance.filters:
+                                keyed_tests[test_key] = {'plat': plat, 'ts': ts}

```

Reply:
see https://github.com/zephyrproject-rtos/zephyr/pull/65834

Reply:
Incorporated #65834's changes.